### DISCUSSION TITLE
How to force set world size

### DISCUSSION THREAD

**Discussion by agunapal** [mpi4py/mpi4py#544] [Q&A]
I am trying to figure out how to force set the world size without using `mpirun -n 4 python test.py`

I want to be able to see 4 being printed when i run `python test.py`. 
the below program prints 1.

Could you please tell me if this can be done?

```
import mpi4py.MPI as MPI
import os

comm = MPI.COMM_WORLD

# Set the world size
os.environ['OMPI_COMM_WORLD_SIZE'] = '4'
os.environ['OMPI_COMM_WORLD_RANK'] = '0'

print(comm.Get_size())
```

**Comment by dalcinl** [comment]
> I am trying to figure out how to force set the world size without using `mpirun -n 4 python test.py`

When you `mpirun -n 4 python ...`, four processes are executed, and after MPI initialization (that in Python happens when you import the MPI module), then COMM_WORLD is set up such that ir reports its size to be 4. If you run `python ...`, then there is a single processes (executed by you), then then when MPI initializes it not able to detect anything else around (because mpirun was not used) and the COMM_WORLD gets a size of 1. Once COMM_WOLD gets its size set upon initialization, there is no way to change it.

> ```
> # Set the world size
> os.environ['OMPI_COMM_WORLD_SIZE'] = '4'
> os.environ['OMPI_COMM_WORLD_RANK'] = '0'
> ```

It is true that `mpirun` sets these environment variables, but you should consider them as informative. Changing them will have no effect on COMM_WORLD.

Long story short, what you are trying to do cannot be done. Moreover, at the risk of sounding offensive, I would further add that what you are asking makes no sense. You did not disclose the true intentions of your questions, I mean, why are you trying to accomplish. If you elaborate a bit more, then maybe I can offer a solution or tip.

**Comment by agunapal** [comment]
Hi @dalcinl Thanks for the reply!

Sorry, didn't give the context as it can be a bit confusing

I am a maintainer of [TorchServe](https://github.com/pytorch/serve) . For multi process communication, we use [torchrun](https://pytorch.org/docs/stable/elastic/run.html) which has worked well till now.

We are trying to integrate trt-llm as a backend into TorchServe. However, trt-llm uses mpirun .
We were trying to see if we could use torchrun args to setup mpirun variables, This would be simpler from a code management/ installation perspective.

Our fallback option is to support both torchrun and mpirun if there is no easy solution.

**Comment by dalcinl** [comment]
I think you only change to get these two things interoperate is trough [PyTorch's MPI backend](https://pytorch.org/docs/stable/distributed.html#backends-that-come-with-pytorch`). That could be  quite restrictive, though. My uninformed guess so far is that you should support both `torchrun` and `mpirun`. What I can tell you for sure is that you will not be able to make the MPI runtime accept being managed by `torchrun`, not at least without `torchrun` somehow exposing to its children process the PMI1/PMI2/PMIx server interfaces MPI libraries usually depend on such that MPI processes can wire-up each other. Simply setting a few environment variables here and there will definitely not make your day.

@leofang Maybe you know anyone that can jump in and provide more proper advice?
